### PR TITLE
Feature/port physics module to new core

### PR DESCRIPTION
**Pull request by Developer-The-Great**

### CODE REVIEW

File: applications/sandbox/systems/physics_test_system.cpp
Comment (LeonBrands) [inline applications/sandbox/systems/physics_test_system.cpp]: remove commented out code

File: applications/sandbox/systems/physics_test_system.cpp
Comment (LeonBrands) [inline applications/sandbox/systems/physics_test_system.cpp]: remove commented out code

File: applications/sandbox/systems/physics_test_system.cpp
Comment (LeonBrands) [inline applications/sandbox/systems/physics_test_system.cpp]: remove commented out code

File: applications/sandbox/systems/physics_test_system.cpp
Comment (LeonBrands) [inline applications/sandbox/systems/physics_test_system.cpp]: remove commented out code

File: applications/sandbox/systems/physics_test_system.cpp
Comment (LeonBrands) [inline applications/sandbox/systems/physics_test_system.cpp:340]: fix double newline

File: applications/sandbox/systems/physics_test_system.cpp
Comment (LeonBrands) [inline applications/sandbox/systems/physics_test_system.cpp]: formatting for these function calls seem weird

File: applications/sandbox/systems/physics_test_system.cpp
Comment (LeonBrands) [inline applications/sandbox/systems/physics_test_system.cpp]: the comments here are a bit excessive, the createQuickhullTestObject function calls are one-liners and the variable name (cubeH, hammerH, etc) is sufficiently descriptive

File: applications/sandbox/systems/physics_test_system.cpp
Comment (LeonBrands) [inline applications/sandbox/systems/physics_test_system.cpp]: move to the PhysicsTestSystem class as private variables

File: applications/sandbox/systems/physics_test_system.cpp
Comment (LeonBrands) [inline applications/sandbox/systems/physics_test_system.cpp]: remove commented out code

File: applications/sandbox/systems/physics_test_system.hpp
Comment (LeonBrands) [inline applications/sandbox/systems/physics_test_system.hpp]: never use `using namespace` in a header

File: legion/engine/physics/halfedgeface.cpp
Comment (LeonBrands) [inline legion/engine/physics/halfedgeface.cpp]: You're comparing integers of different signedness here, keep in mind that i is of type size_t and thus (i -1) will result in UINT_MAX when it overflows. This particular comparison might work out because the -1 will be implicitly converted to uint and overflow as well, but this assumption that i is capable of having negative numbers is error-prone. Either convert i to int or use -1u for the comparison so that the overflow is intentional

File: legion/engine/physics/halfedgeface.cpp
Comment (LeonBrands) [inline legion/engine/physics/halfedgeface.cpp]: remove excessive newlines

File: legion/engine/physics/halfedgeface.cpp
Comment (LeonBrands) [inline legion/engine/physics/halfedgeface.cpp]: remove excessive newline

File: legion/engine/physics/physics_statics.hpp
Comment (LeonBrands) [inline legion/engine/physics/physics_statics.hpp]: remove commented out code

**Review by LeonBrands — CHANGES_REQUESTED** [review]

File: applications/sandbox/systems/physics_test_system.hpp
Comment (GlynLeine) [reply applications/sandbox/systems/physics_test_system.hpp]: i'd suggest using the `lgn` namespace alias for everything that would require the namespace without this line

File: legion/engine/physics/colliders/physicscollider.hpp
Comment (GlynLeine) [inline legion/engine/physics/colliders/physicscollider.hpp]: are you trying to move the unique pointers out? i don't think this is legal code?

File: legion/engine/physics/data/collider_face_to_vert.cpp
Comment (GlynLeine) [inline legion/engine/physics/data/collider_face_to_vert.cpp]: nitpick, but i generally like keeping a line of whitespace between fucntions

File: legion/engine/physics/data/collider_face_to_vert.hpp
Comment (GlynLeine) [inline legion/engine/physics/data/collider_face_to_vert.hpp]: technically if this is a non owning pointer you could use the `core::pointer<T>` wrapper, it doesn't allow you to call new or delete on it. but that's also personal preference. idm keeping it like this either

File: legion/engine/physics/data/convexconvexpenetrationquery.cpp
Comment (GlynLeine) [inline legion/engine/physics/data/convexconvexpenetrationquery.cpp]: have you checked if this line ending change means it's not `LF` instead of `CRLF`?

File: legion/engine/physics/data/physics_manifold.hpp
Comment (GlynLeine) [inline legion/engine/physics/data/physics_manifold.hpp:38]: commented out code under this line

File: legion/engine/physics/data/physics_manifold.hpp
Comment (GlynLeine) [inline legion/engine/physics/data/physics_manifold.hpp]: psst, you can use the `->` operator on the new component handles

File: legion/engine/physics/data/physics_manifold_precursor.hpp
Comment (GlynLeine) [inline legion/engine/physics/data/physics_manifold_precursor.hpp:20]: it's generally not recommended to store component pointers, they are not guaranteed to stay valid between frames. but this is fine if this never persists between frames. but in that case maybe add a comment?

File: legion/engine/physics/data/physics_manifold.hpp
Comment (GlynLeine) [inline legion/engine/physics/data/physics_manifold.hpp:35]: it's generally not recommended to store component pointers, they are not guaranteed to stay valid between frames. but this is fine if this never persists between frames. but in that case maybe add a comment?

File: legion/engine/physics/halfedgeedge.cpp
Comment (GlynLeine) [inline legion/engine/physics/halfedgeedge.cpp]: `debug::user_projectDrawLine` is not guaranteed to exist, use `debug::drawLine()`. `debug::user_projectDrawLine` won't exist without sandbox. also in the other instances of the function call in this file

File: legion/engine/physics/halfedgeface.cpp
Comment (GlynLeine) [reply legion/engine/physics/halfedgeface.cpp]: I'd even just say change it to: `((i  == 0ull ? (edges.size() - 1ull) : (i - 1ull));`

File: legion/engine/physics/physics_statics.cpp
Comment (GlynLeine) [inline legion/engine/physics/physics_statics.cpp]: you're doing a signed unsigned comparison, and `vertexIndex` gets implictly casted to a `size_type` every iteration. this is not costless.

File: legion/engine/physics/physics_statics.cpp
Comment (GlynLeine) [inline legion/engine/physics/physics_statics.cpp]: I have a feeling this probably should be in a more public place than locally inside this function? also local static variables aren't 0 cost. they have a thread-safe initialization check around them that will be called every time this function is called.

File: legion/engine/physics/physics_statics.cpp
Comment (GlynLeine) [inline legion/engine/physics/physics_statics.cpp]: same as before, signed unsigned compare, and `at(size_t)` will cause a cast to `size_type`

File: legion/engine/physics/physics_statics.cpp
Comment (GlynLeine) [inline legion/engine/physics/physics_statics.cpp:769]: at some point, maybe with the refactor, this should probably get properly handled using `common::result` instead of crashing the entire engine

File: legion/engine/physics/physics_statics.cpp
Comment (GlynLeine) [inline legion/engine/physics/physics_statics.cpp:975]: same as mentioned before about `common::result`

File: legion/engine/physics/systems/physicssystem.cpp
Comment (GlynLeine) [inline legion/engine/physics/systems/physicssystem.cpp]: wait why exactly did we stop using the fixed interval process?

File: legion/engine/physics/systems/physicssystem.hpp
Comment (GlynLeine) [inline legion/engine/physics/systems/physicssystem.hpp]: fyi, these all work:
```cpp
rigidbody& rb = rigidbodies[async::this_job::get_id()];
```
or
```cpp
auto& rb = *rigidbodies[async::this_job::get_id()];
```
or
```cpp
auto& rb = rigidbodies[async::this_job::get_id()].get();
```

File: legion/engine/physics/colliders/physicscollider.hpp
Comment (Developer-The-Great) [reply legion/engine/physics/colliders/physicscollider.hpp]: there is no move being done here, some of the floats of  Physics_Contact are being copied to ConvergenceIdentifier

Comment (GlynLeine) [reply legion/engine/physics/colliders/physicscollider.hpp]: this this should still be an `auto&` since i don't recall calling functions or writing data to a rvalue reference being generally considered well defined. since rvalues are meant to be temporary values. (right hand side values)

Comment (Developer-The-Great) [reply legion/engine/physics/colliders/physicscollider.hpp]: ah did not notice the '&&' will fix that

File: legion/engine/physics/data/physics_manifold.hpp
Comment (Developer-The-Great) [reply legion/engine/physics/data/physics_manifold.hpp:35]: This doesn't persist between frames, will add that comment mentioning that

File: legion/engine/physics/data/physics_manifold_precursor.hpp
Comment (Developer-The-Great) [reply legion/engine/physics/data/physics_manifold_precursor.hpp:20]: this never persist between frames, will add that comment

File: legion/engine/physics/halfedgeface.cpp
Comment (Developer-The-Great) [reply legion/engine/physics/halfedgeface.cpp]: I'll convert it i to int since it would be much more readable

Comment (GlynLeine) [reply legion/engine/physics/halfedgeface.cpp]: actually that was the wrong thing to take away here, since `.size()` will always return a `size_type` anyways. turning i into an int will just cause signed unsigned comparison, and implicit casts from `size_type` to `int`. actually `nextIndex` and `prevIndex` should also be `size_type`.

generally if you're dealing with indices and sizes and amounts, you want to be using `size_type`

File: legion/engine/physics/physics_statics.cpp
Comment (Developer-The-Great) [reply legion/engine/physics/physics_statics.cpp]: i do recall that I wanted this to be a constexpr value in physics::constants, but this wasn't possible because there is no constexpr version of math::cos. I could just declare the literal value of math::cos(math::deg2rad(3.0f)) in physics::constants, is that sufficient?

Comment (GlynLeine) [reply legion/engine/physics/physics_statics.cpp]: yeah if you can't make it `constexpr`, it's fine to make it `static const`. it won't be generated on compile time, but it will be generated at startup.

File: legion/engine/physics/data/collider_face_to_vert.hpp
Comment (Developer-The-Great) [reply legion/engine/physics/data/collider_face_to_vert.hpp]: image

core::pointer currently doesn't have a constructor that takes in the T* in question. Which means I cant value initialize my pointers to nullptr. Will keep using raw pointers in this specific situation for now.

File: legion/engine/physics/physics_statics.cpp
Comment (Developer-The-Great) [reply legion/engine/physics/physics_statics.cpp:769]: would you prefer me to handle this in this pr or in another separate pr?

File: legion/engine/physics/systems/physicssystem.cpp
Comment (Developer-The-Great) [reply legion/engine/physics/systems/physicssystem.cpp]: Initially, this was because there seemed to be something wrong with createProcess causing it to call fixedUpdate in the wrong intervals. I looked at this section of the code again and did some tests, apparently, it was something entirely unrelated (my bad!). But yeah good catch, will revert this.

File: legion/engine/physics/systems/physicssystem.hpp
Comment (Developer-The-Great) [reply legion/engine/physics/systems/physicssystem.hpp]: good to know!

File: legion/engine/physics/data/collider_face_to_vert.hpp
Comment (GlynLeine) [reply legion/engine/physics/data/collider_face_to_vert.hpp]: that's becasue it's a trivial type:
```cpp
core::pointer<HalfEdgeFace> face2 = { nullptr };
```

File: legion/engine/physics/physics_statics.cpp
Comment (GlynLeine) [reply legion/engine/physics/physics_statics.cpp:769]: this will probably require a bigger architectural change, just like the use of raw memory management throughout physics. I think this should come in with whatever PR will come with that one.

File: legion/engine/physics/systems/physicssystem.cpp
Comment (GlynLeine) [reply legion/engine/physics/systems/physicssystem.cpp]: If you would like to change the behaviour of fixed interval process, please let me know. physics was the main motivation behind it's existence. (and also the only that uses it rn) So it's fine to say add the maximum updates per frame that you added here.

Comment (GlynLeine) [reply legion/engine/physics/systems/physicssystem.cpp]: or alternatively you could ofc add it yourself

Comment (Developer-The-Great) [reply legion/engine/physics/systems/physicssystem.cpp]: I might change my mind about this later on, but I think I prefer using createProcess since this essentially achieves what we want to do (update physics in a fixed interval) with comparitively less code.

**Review by LeonBrands — APPROVED** [review]
changes lgtm. I'd wait for a review from Glyn before merging though

File: legion/engine/physics/systems/physicssystem.cpp
Comment (GlynLeine) [reply legion/engine/physics/systems/physicssystem.cpp]: no what i meant is: you're allowed to go into system and scheduling to change how fixed interval process works. admittedly i didn't spend all too much research into how it should or shouldn't behave.

Comment (Developer-The-Great) [reply legion/engine/physics/systems/physicssystem.cpp]: alright, but this is not something I need to do right now since it works as intended for now

File: applications/sandbox/systems/physics_test_system.cpp
Comment (GlynLeine) [inline applications/sandbox/systems/physics_test_system.cpp:111]: I get that this is just a test system, so i don't rly care tbh. but keep in mind that any of these handles might be invalid. and that will cause an exception.

File: applications/sandbox/systems/physics_test_system.cpp
Comment (GlynLeine) [inline applications/sandbox/systems/physics_test_system.cpp:205]: tbh i would use auto& here, but up to you XD

File: applications/sandbox/systems/physics_test_system.cpp
Comment (GlynLeine) [inline applications/sandbox/systems/physics_test_system.cpp:291]: yeah i'd recommend getting used to using lot's of auto, but ig this is prolly pretty old code since i do see you using auto everywhere else XD

**Review by GlynLeine — APPROVED** [review]
lgtm!

**Pull request merged**
